File: .agents/skills/search-israel-trails/scripts/search_trails.py
#!/usr/bin/env python3
"""Three-API trail search for Israel Hiking Map.

Usage:
    python search_trails.py <query> [--max N] [--language en|he]

Exits 0 and prints JSON array of trail objects on success.
Exits 1 and prints {"error": "..."} on failure.
"""

from __future__ import annotations

import json
import math
import sys
import time
import urllib.parse
import urllib.request
import urllib.error

DEFAULT_MAX = 3
DEFAULT_MAX_KM = 30  # flag routes longer than this as long_distance_route
OVERPASS_URL = "https://overpass-api.de/api/interpreter"
IHM_BASE = "https://israelhiking.osm.org.il/api"
TIMEOUT = 20
ROUTING_TIMEOUT = 10   # routing API is fast; fail quickly and fall back
ROUTING_MIN_GAP = 0.5  # seconds between routing calls — respect IHM rate limit


# ── HTTP helpers ─────────────────────────────────────────────────────────────

def get_json(url: str) -> object:
    with urllib.request.urlopen(url, timeout=TIMEOUT) as r:
        return json.loads(r.read().decode())


def post_json(url: str, body: str) -> object:
    # Overpass API expects form-encoded body: data=<query>
    encoded = urllib.parse.urlencode({"data": body}).encode()
    req = urllib.request.Request(url, data=encoded, method="POST")
    req.add_header("Content-Type", "application/x-www-form-urlencoded")
    req.add_header("User-Agent", "TrailMate/1.0 (travel planning)")
    with urllib.request.urlopen(req, timeout=TIMEOUT) as r:
        return json.loads(r.read().decode())


# ── IHM Routing API ──────────────────────────────────────────────────────────

def routing_distance_and_elevation(
    start: tuple[float, float],
    end: tuple[float, float],
) -> dict | None:
    """Call IHM's routing API between two WGS84 points and return accurate metrics.

    Returns a dict with ``distance_km``, ``elevation_gain_m``, ``elevation_loss_m``
    on success, or ``None`` if the call fails or the route is degenerate.

    The routing engine follows actual hiking paths, so the distance is the
    real walking distance — not the sum of unsorted geometry segments.
    """
    # Skip routing when start ≈ end (loop trail whose geometry gave us the
    # same first/last point) — the router would return near-zero distance.
    if haversine(start, end) < 200:
        return None

    url = (
        f"{IHM_BASE}/routing"
        f"?from={start[0]},{start[1]}"
        f"&to={end[0]},{end[1]}"
        f"&type=Hike"
    )
    try:
        with urllib.request.urlopen(url, timeout=ROUTING_TIMEOUT) as r:
            data = json.loads(r.read().decode())
    except Exception:
        return None

    features = data.get("features", [])
    if not features:
        return None

    # Coordinates are [lng, lat, elevation]
    coords = features[0].get("geometry", {}).get("coordinates", [])
    if len(coords) < 2:
        return None

    # Distance: sum consecutive haversine distances on the ROUTED path
    dist_m = sum(
        haversine((coords[i][1], coords[i][0]), (coords[i + 1][1], coords[i + 1][0]))
        for i in range(len(coords) - 1)
    )
    dist_km = round(dist_m / 1000, 1)
    if dist_km == 0:
        return None

    # Elevation: cumulative gain and loss from per-point elevations
    elevations = [c[2] for c in coords if len(c) > 2 and c[2] is not None]
    gain = loss = 0
    for i in range(1, len(elevations)):
        diff = elevations[i] - elevations[i - 1]
        if diff > 0:
            gain += diff
        else:
            loss += abs(diff)

    return {
        "distance_km": dist_km,
        "elevation_gain_m": round(gain),
        "elevation_loss_m": round(loss),
    }


# ── Step 1: IHM search ───────────────────────────────────────────────────────

def ihm_search(query: str, language: str, max_results: int) -> list[dict]:
    encoded = urllib.parse.quote(query)
    url = f"{IHM_BASE}/search/{encoded}?language={language}"
    results = get_json(url)
    # Only keep results that are OSM relations — these are named trail routes.
    # way_ results are often short footpaths or streets; node_ results are
    # POIs/localities that can't be enriched with geometry or trail tags.
    return [
        r for r in results
        if "hike" in r.get("icon", "") and r.get("id", "").startswith("relation_")
    ][:max_results]


# ── Step 2: Overpass tag enrichment ──────────────────────────────────────────

def overpass_tags(osm_type: str, osm_id: str) -> dict:
    """Fetch OSM tags for a relation or way."""
    if osm_type == "relation":
        data = f"[out:json];relation({osm_id});out tags;"
    else:
        data = f"[out:json];way({osm_id});out tags;"
    resp = post_json(OVERPASS_URL, data)
    elements = resp.get("elements", [])
    return elements[0].get("tags", {}) if elements else {}


def parse_color(osmc_symbol: str) -> str:
    known = {"red", "blue", "green", "black", "orange", "white", "yellow"}
    for part in osmc_symbol.split(":"):
        word = part.split("_")[0]
        if word in known:
            return word
    return ""


# ── Step 3a: Overpass geometry ────────────────────────────────────────────────

def overpass_geometry(osm_type: str, osm_id: str) -> list[tuple[float, float]]:
    """Fetch ordered coordinates for a relation or way."""
    coords = []
    if osm_type == "relation":
        data = f"[out:json];relation({osm_id});way(r);out geom;"
        resp = post_json(OVERPASS_URL, data)
        for way in resp.get("elements", []):
            for node in way.get("geometry", []):
                coords.append((node["lat"], node["lon"]))
    else:
        data = f"[out:json];way({osm_id});out geom;"
        resp = post_json(OVERPASS_URL, data)
        for element in resp.get("elements", []):
            for node in element.get("geometry", []):
                coords.append((node["lat"], node["lon"]))
    return coords


# ── Step 3b: Haversine distance ───────────────────────────────────────────────

def haversine(p1: tuple[float, float], p2: tuple[float, float]) -> float:
    R = 6_371_000
    lat1, lon1 = math.radians(p1[0]), math.radians(p1[1])
    lat2, lon2 = math.radians(p2[0]), math.radians(p2[1])
    a = math.sin((lat2 - lat1) / 2) ** 2 + math.cos(lat1) * math.cos(lat2) * math.sin((lon2 - lon1) / 2) ** 2
    return R * 2 * math.asin(math.sqrt(a))


def total_distance_km(coords: list[tuple[float, float]]) -> float:
    total = sum(haversine(coords[i], coords[i + 1]) for i in range(len(coords) - 1))
    return round(total / 1000, 1)


# ── Step 3c: IHM elevation ────────────────────────────────────────────────────

def elevation_profile(coords: list[tuple[float, float]]) -> tuple[float, float]:
    """Return (gain_m, loss_m) from sampled elevation data."""
    step = max(1, len(coords) // 20)
    sample = coords[::step][:20]
    points_param = "|".join(f"{lat},{lon}" for lat, lon in sample)
    url = f"{IHM_BASE}/elevation?points={urllib.parse.quote(points_param)}"
    elevs = get_json(url)
    gain = round(sum(max(0.0, elevs[i + 1] - elevs[i]) for i in range(len(elevs) - 1)))
    loss = round(sum(max(0.0, elevs[i] - elevs[i + 1]) for i in range(len(elevs) - 1)))
    return gain, loss


# ── Step 3d: Difficulty ───────────────────────────────────────────────────────

def classify_difficulty(distance_km: float, gain_m: float) -> str:
    score = distance_km + gain_m / 100
    if score < 5:
        return "easy"
    elif score < 15:
        return "moderate"
    return "hard"


# ── Step 3e: Duration estimate (Naismith's rule) ─────────────────────────────

def estimate_duration(distance_km: float, gain_m: float) -> str:
    """Naismith: 4 km/h walking pace + 1 h per 600 m ascent."""
    hours_raw = distance_km / 4.0 + gain_m / 600.0
    # Round to nearest half-hour
    half_hours = round(hours_raw * 2)
    hours = half_hours // 2
    mins = 30 if half_hours % 2 else 0
    if hours == 0:
        return f"{mins or 30} min"
    if mins:
        return f"{hours}h {mins}min"
    return f"{hours}h"


# ── Step 3f: Car logistics ────────────────────────────────────────────────────

def car_logistics(tags: dict, coords: list[tuple[float, float]]) -> str:
    """Determine whether the trail is a loop (1 car) or linear (2 cars/shuttle)."""
    roundtrip = tags.get("roundtrip", "").lower()
    if roundtrip == "yes":
        return "loop — 1 car"
    if roundtrip == "no":
        return "linear — 2 cars or shuttle"
    # Heuristic: if first and last coordinate are within 500 m, treat as loop
    if len(coords) >= 2 and haversine(coords[0], coords[-1]) < 500:
        return "loop — 1 car"
    if coords:
        return "linear — 2 cars or shuttle"
    return "unknown"


# ── Main flow ─────────────────────────────────────────────────────────────────

def enrich(trail: dict) -> dict:
    info: dict = {
        "name": trail.get("title"),
        "display_name": trail.get("displayName"),
        "location": trail.get("location"),
    }

    raw_id = trail.get("id", "")
    if raw_id.startswith("relation_"):
        osm_type, osm_id = "relation", raw_id.replace("relation_", "")
    elif raw_id.startswith("way_"):
        osm_type, osm_id = "way", raw_id.replace("way_", "")
    else:
        return info  # node or unknown — not enrichable via Overpass

    tags: dict = {}

    # Step 2: tags
    try:
        tags = overpass_tags(osm_type, osm_id)
        color = parse_color(tags.get("osmc:symbol", ""))
        if color:
            info["trail_color"] = color
        network_map = {"lwn": "local", "rwn": "regional", "nwn": "national"}
        if tags.get("network"):
            info["network"] = network_map.get(tags["network"], tags["network"])
        if tags.get("ref"):
            info["ref"] = tags["ref"]
        if tags.get("description"):
            info["description"] = tags["description"]
        # Start / end named locations
        if tags.get("from"):
            info["trailhead_from"] = tags["from"]
        if tags.get("to"):
            info["trailhead_to"] = tags["to"]
        # Distance from tag
        if tags.get("distance"):
            try:
                info["distance_km"] = float(tags["distance"].replace("km", "").strip())
            except ValueError:
                pass
        # Elevation from tags (more reliable than computed when present)
        if tags.get("ascent"):
            try:
                info["elevation_gain_m"] = int(tags["ascent"])
            except ValueError:
                pass
        if tags.get("descent"):
            try:
                info["elevation_loss_m"] = int(tags["descent"])
            except ValueError:
                pass
        if tags.get("operator"):
            info["operator"] = tags["operator"]
        if tags.get("website") or tags.get("url"):
            info["website"] = tags.get("website") or tags.get("url")
    except Exception:
        pass

    # Step 3: geometry — used only for trailhead coords and loop detection.
    coords: list[tuple[float, float]] = []
    try:
        coords = overpass_geometry(osm_type, osm_id)
        if coords:
            info["trailhead_coords"] = {"lat": round(coords[0][0], 6), "lng": round(coords[0][1], 6)}
    except Exception:
        pass

    # Step 4: accurate distance + elevation via IHM routing API.
    # Routing follows actual hiking paths, so it avoids the inflated sums
    # produced by concatenating unsorted OSM relation member ways.
    # Only runs when tags didn't already supply a distance value.
    routed = None
    if "distance_km" not in info and len(coords) >= 2:
        try:
            time.sleep(ROUTING_MIN_GAP)
            routed = routing_distance_and_elevation(coords[0], coords[-1])
        except Exception:
            pass

    if routed:
        info["distance_km"]     = routed["distance_km"]
        info["elevation_gain_m"] = routed["elevation_gain_m"]
        info["elevation_loss_m"] = routed["elevation_loss_m"]
    else:
        # Fallback: geometry-based calculation (less accurate for complex relations)
        if "distance_km" not in info and len(coords) >= 2:
            info["distance_km"] = total_distance_km(coords)
        if ("elevation_gain_m" not in info or "elevation_loss_m" not in info) and len(coords) >= 2:
            try:
                gain_fb, loss_fb = elevation_profile(coords)
                if "elevation_gain_m" not in info:
                    info["elevation_gain_m"] = gain_fb
                if "elevation_loss_m" not in info:
                    info["elevation_loss_m"] = loss_fb
            except Exception:
                pass

    # Derived fields
    dist = info.get("distance_km", 0)
    gain = info.get("elevation_gain_m", 0)

    if dist:
        if "difficulty" not in info:
            info["difficulty"] = classify_difficulty(dist, gain)
        info["estimated_duration"] = estimate_duration(dist, gain)

    info["car_logistics"] = car_logistics(tags, coords)

    return info


def search(
    query: str,
    language: str = "en",
    max_results: int = DEFAULT_MAX,
    max_km: float = DEFAULT_MAX_KM,
) -> list[dict]:
    trails = ihm_search(query, language, max_results)
    if not trails:
        return []
    enriched = [enrich(t) for t in trails]
    for t in enriched:
        dist = t.get("distance_km", 0)
        if dist > max_km:
            # OSM geometry for a regional/national trail relation sums ALL member
            # ways regardless of order, producing inflated distances (e.g. 481 km).
            # Flag it so callers know to suppress the numbers and rely on tiuli.
            t["long_distance_route"] = True
            t.pop("distance_km", None)
            t.pop("estimated_duration", None)
    return enriched


# ── CLI entry point ───────────────────────────────────────────────────────────

if __name__ == "__main__":
    args = sys.argv[1:]
    if not args:
        print(json.dumps({"error": "Usage: search_trails.py <query> [--max N] [--language en|he]"}))
        sys.exit(1)

    query = args[0]
    max_results = DEFAULT_MAX
    max_km = DEFAULT_MAX_KM
    language = "en"

    i = 1
    while i < len(args):
        if args[i] == "--max" and i + 1 < len(args):
            max_results = min(int(args[i + 1]), 5)
            i += 2
        elif args[i] == "--max-km" and i + 1 < len(args):
            max_km = float(args[i + 1])
            i += 2
        elif args[i] == "--language" and i + 1 < len(args):
            language = args[i + 1]
            i += 2
        else:
            i += 1

    try:
        results = search(query, language, max_results, max_km)
        print(json.dumps(results, ensure_ascii=False, indent=2))
    except Exception as e:
        print(json.dumps({"error": str(e)}))
        sys.exit(1)

File: web/lib/tools/searchTrails.ts
// searchTrails — Israel Hiking Map 3-API enrichment. Ported from
// .agents/skills/search-israel-trails/scripts/search_trails.py.
//
// Secondary geographic trail search (the tiuli catalog is primary). Returns
// real OSM trail routes with distance, elevation, difficulty, and duration.
// No API key required.

import type { ToolDef } from "./types";

const DEFAULT_MAX = 3;
const DEFAULT_MAX_KM = 30;
const OVERPASS_URL = "https://overpass-api.de/api/interpreter";
const IHM_BASE = "https://israelhiking.osm.org.il/api";
const ROUTING_MIN_GAP = 500; // ms between routing calls
const UA = { "User-Agent": "TrailMate/1.0 (travel planning)" };

type Pt = [number, number]; // [lat, lng]

async function getJson(url: string): Promise<any> {
  const res = await fetch(url, { headers: UA });
  if (!res.ok) throw new Error(`HTTP ${res.status}`);
  return res.json();
}

async function postOverpass(body: string): Promise<any> {
  const res = await fetch(OVERPASS_URL, {
    method: "POST",
    headers: { ...UA, "Content-Type": "application/x-www-form-urlencoded" },
    body: new URLSearchParams({ data: body }).toString(),
  });
  if (!res.ok) throw new Error(`HTTP ${res.status}`);
  return res.json();
}

function haversine(p1: Pt, p2: Pt): number {
  const R = 6_371_000;
  const toRad = (d: number) => (d * Math.PI) / 180;
  const lat1 = toRad(p1[0]);
  const lat2 = toRad(p2[0]);
  const a =
    Math.sin((lat2 - lat1) / 2) ** 2 +
    Math.cos(lat1) * Math.cos(lat2) * Math.sin(toRad(p2[1] - p1[1]) / 2) ** 2;
  return R * 2 * Math.asin(Math.sqrt(a));
}

function totalDistanceKm(coords: Pt[]): number {
  let total = 0;
  for (let i = 0; i < coords.length - 1; i++) total += haversine(coords[i], coords[i + 1]);
  return Math.round((total / 1000) * 10) / 10;
}

function classifyDifficulty(distKm: number, gainM: number): string {
  const score = distKm + gainM / 100;
  if (score < 5) return "easy";
  if (score < 15) return "moderate";
  return "hard";
}

function estimateDuration(distKm: number, gainM: number): string {
  const hoursRaw = distKm / 4.0 + gainM / 600.0;
  const halfHours = Math.round(hoursRaw * 2);
  const hours = Math.floor(halfHours / 2);
  const mins = halfHours % 2 ? 30 : 0;
  if (hours === 0) return `${mins || 30} min`;
  return mins ? `${hours}h ${mins}min` : `${hours}h`;
}

function parseColor(osmc: string): string {
  const known = new Set(["red", "blue", "green", "black", "orange", "white", "yellow"]);
  for (const part of osmc.split(":")) {
    const word = part.split("_")[0];
    if (known.has(word)) return word;
  }
  return "";
}

async function ihmSearch(query: string, language: string, maxResults: number): Promise<any[]> {
  const url = `${IHM_BASE}/search/${encodeURIComponent(query)}?language=${language}`;
  const results = await getJson(url);
  return (Array.isArray(results) ? results : [])
    .filter((r) => String(r.icon ?? "").includes("hike") && String(r.id ?? "").startsWith("relation_"))
    .slice(0, maxResults);
}

async function overpassTags(osmType: string, osmId: string): Promise<Record<string, string>> {
  const data =
    osmType === "relation"
      ? `[out:json];relation(${osmId});out tags;`
      : `[out:json];way(${osmId});out tags;`;
  const resp = await postOverpass(data);
  const els = resp.elements ?? [];
  return els[0]?.tags ?? {};
}

async function overpassGeometry(osmType: string, osmId: string): Promise<Pt[]> {
  const coords: Pt[] = [];
  const data =
    osmType === "relation"
      ? `[out:json];relation(${osmId});way(r);out geom;`
      : `[out:json];way(${osmId});out geom;`;
  const resp = await postOverpass(data);
  for (const el of resp.elements ?? []) {
    for (const node of el.geometry ?? []) coords.push([node.lat, node.lon]);
  }
  return coords;
}

async function routingDistanceElevation(start: Pt, end: Pt) {
  if (haversine(start, end) < 200) return null;
  const url = `${IHM_BASE}/routing?from=${start[0]},${start[1]}&to=${end[0]},${end[1]}&type=Hike`;
  let data: any;
  try {
    data = await getJson(url);
  } catch {
    return null;
  }
  const coords = data?.features?.[0]?.geometry?.coordinates ?? [];
  if (coords.length < 2) return null;
  let distM = 0;
  for (let i = 0; i < coords.length - 1; i++) {
    distM += haversine([coords[i][1], coords[i][0]], [coords[i + 1][1], coords[i + 1][0]]);
  }
  const distKm = Math.round((distM / 1000) * 10) / 10;
  if (distKm === 0) return null;
  const elevs = coords.filter((c: number[]) => c.length > 2 && c[2] != null).map((c: number[]) => c[2]);
  let gain = 0;
  let loss = 0;
  for (let i = 1; i < elevs.length; i++) {
    const diff = elevs[i] - elevs[i - 1];
    if (diff > 0) gain += diff;
    else loss += Math.abs(diff);
  }
  return { distance_km: distKm, elevation_gain_m: Math.round(gain), elevation_loss_m: Math.round(loss) };
}

function carLogistics(tags: Record<string, string>, coords: Pt[]): string {
  const roundtrip = String(tags.roundtrip ?? "").toLowerCase();
  if (roundtrip === "yes") return "loop — 1 car";
  if (roundtrip === "no") return "linear — 2 cars or shuttle";
  if (coords.length >= 2 && haversine(coords[0], coords[coords.length - 1]) < 500) return "loop — 1 car";
  return coords.length ? "linear — 2 cars or shuttle" : "unknown";
}

async function enrich(trail: any): Promise<Record<string, any>> {
  const info: Record<string, any> = {
    name: trail.title,
    display_name: trail.displayName,
    location: trail.location,
  };
  const rawId: string = trail.id ?? "";
  let osmType: string;
  let osmId: string;
  if (rawId.startsWith("relation_")) [osmType, osmId] = ["relation", rawId.replace("relation_", "")];
  else if (rawId.startsWith("way_")) [osmType, osmId] = ["way", rawId.replace("way_", "")];
  else return info;

  let tags: Record<string, string> = {};
  try {
    tags = await overpassTags(osmType, osmId);
    const color = parseColor(tags["osmc:symbol"] ?? "");
    if (color) info.trail_color = color;
    const networkMap: Record<string, string> = { lwn: "local", rwn: "regional", nwn: "national" };
    if (tags.network) info.network = networkMap[tags.network] ?? tags.network;
    if (tags.ref) info.ref = tags.ref;
    if (tags.description) info.description = tags.description;
    if (tags.from) info.trailhead_from = tags.from;
    if (tags.to) info.trailhead_to = tags.to;
    if (tags.distance) {
      const d = parseFloat(tags.distance.replace("km", "").trim());
      if (!Number.isNaN(d)) info.distance_km = d;
    }
    if (tags.ascent && !Number.isNaN(parseInt(tags.ascent))) info.elevation_gain_m = parseInt(tags.ascent);
    if (tags.descent && !Number.isNaN(parseInt(tags.descent))) info.elevation_loss_m = parseInt(tags.descent);
    if (tags.operator) info.operator = tags.operator;
    if (tags.website || tags.url) info.website = tags.website || tags.url;
  } catch {
    /* tags best-effort */
  }

  let coords: Pt[] = [];
  try {
    coords = await overpassGeometry(osmType, osmId);
    if (coords.length) {
      info.trailhead_coords = {
        lat: Math.round(coords[0][0] * 1e6) / 1e6,
        lng: Math.round(coords[0][1] * 1e6) / 1e6,
      };
      info.start_maps = `https://www.google.com/maps?q=${coords[0][0]},${coords[0][1]}`;
    }
  } catch {
    /* geometry best-effort */
  }

  let routed = null;
  if (!("distance_km" in info) && coords.length >= 2) {
    try {
      await new Promise((r) => setTimeout(r, ROUTING_MIN_GAP));
      routed = await routingDistanceElevation(coords[0], coords[coords.length - 1]);
    } catch {
      /* fall back below */
    }
  }
  if (routed) {
    info.distance_km = routed.distance_km;
    info.elevation_gain_m = routed.elevation_gain_m;
    info.elevation_loss_m = routed.elevation_loss_m;
  } else if (!("distance_km" in info) && coords.length >= 2) {
    info.distance_km = totalDistanceKm(coords);
  }

  const dist = info.distance_km ?? 0;
  const gain = info.elevation_gain_m ?? 0;
  if (dist) {
    if (!("difficulty" in info)) info.difficulty = classifyDifficulty(dist, gain);
    info.estimated_duration = estimateDuration(dist, gain);
  }
  info.car_logistics = carLogistics(tags, coords);
  return info;
}

export const searchTrails: ToolDef = {
  schema: {
    type: "function",
    function: {
      name: "search_trails",
      description:
        "Secondary geographic trail search via the Israel Hiking Map / OpenStreetMap. " +
        "Use when the tiuli catalog (search_tiuli) has no good match for an area, or " +
        "to get computed distance, elevation, and difficulty for a named OSM trail.",
      parameters: {
        type: "object",
        properties: {
          query: { type: "string", description: "Trail or area name, e.g. 'Yehudiya', 'Arbel'." },
          max: { type: "integer", description: "Max trails (default 3, cap 5)." },
          language: { type: "string", enum: ["en", "he"], description: "Search language. Default en." },
        },
        required: ["query"],
      },
    },
  },
  async execute(args: Record<string, any>) {
    const query = String(args.query ?? "").trim();
    if (!query) return { status: "error", message: "query is required" };
    const maxResults = Math.min(Number(args.max ?? DEFAULT_MAX), 5);
    const language = args.language === "he" ? "he" : "en";
    try {
      const trails = await ihmSearch(query, language, maxResults);
      if (trails.length === 0) return { status: "success", trails: [] };
      const enriched = await Promise.all(trails.map((t) => enrich(t)));
      for (const t of enriched) {
        if ((t.distance_km ?? 0) > DEFAULT_MAX_KM) {
          // OSM relation geometry sums all member ways unordered → inflated
          // distance. Flag and drop the numbers; rely on tiuli instead.
          t.long_distance_route = true;
          delete t.distance_km;
          delete t.estimated_duration;
        }
      }
      return { status: "success", trails: enriched };
    } catch (e: any) {
      return { status: "error", message: String(e?.message ?? e) };
    }
  },
};
